### PR TITLE
CORE-17656 - Upgrade testing for network

### PR DESCRIPTION
**Pull request by nikinagy**
Added new function to be able to re-register when we don't have the onboarding metadata locally.
This PR contains a fix for MGM re-registration: we should let the MGM modify their platform and software version to support the upgrades.

Pipeline build:
https://ci02.dev.r3.com/blue/organizations/jenkins/Corda5%2Fcorda5-upgrade-tests/detail/PR-423/12/tests/

### CODE REVIEW

File: components/membership/registration-impl/src/main/kotlin/net/corda/membership/impl/registration/dynamic/mgm/MGMRegistrationContextValidator.kt [hunk 1]
```diff
@@ -96,7 +96,9 @@ internal class MGMRegistrationContextValidator(
         val lastMemberContext = lastMemberInfo.memberProvidedContext.toMap()
         val diff = ((newMemberContext.entries - lastMemberContext.entries) + (lastMemberContext.entries - newMemberContext.entries))
             .filterNot {
-                it.key.startsWith(MemberInfoExtension.ENDPOINTS)
+                it.key.startsWith(MemberInfoExtension.ENDPOINTS) ||
```
Comment (yift-r3) [inline components/membership/registration-impl/src/main/kotlin/net/corda/membership/impl/registration/dynamic/mgm/MGMRegistrationContextValidator.kt:99]: Should we put changes to the production code in a testing pull request?

File: testing/e2e-test-utilities/src/main/kotlin/net/corda/e2etest/utilities/MembershipUtils.kt [hunk 2]
```diff
@@ -144,6 +144,37 @@ fun NetworkOnboardingMetadata.reregisterMember(
...
+ * Should be mainly used in upgrade scenarios where we cannot re-use the onboarding metadata.
+ */
+fun ClusterInfo.reregisterMember(
+    holdingId: String,
+    x500Name: String,
+    registrationContext: Map<String, String>,
+    contextToMerge: Map<String, String?> = emptyMap(),
+    waitForApproval: Boolean = true
+): NetworkOnboardingMetadata {
+    val newContext = registrationContext.toMutableMap()
```
Comment (yift-r3) [inline testing/e2e-test-utilities/src/main/kotlin/net/corda/e2etest/utilities/MembershipUtils.kt]: Maybe something like:
```
    val newContext = registrationContext.mapNotNull { (key, value) ->
        if (contextToMerge.containsKey(key)) {
            contextToMerge[key]?.let {
                key to it
            }
        } else {
            key to value
        }
    }.toMap()
```?

File: components/membership/registration-impl/src/main/kotlin/net/corda/membership/impl/registration/dynamic/mgm/MGMRegistrationContextValidator.kt (hunk 1 above)
Comment (nikinagy) [reply components/membership/registration-impl/src/main/kotlin/net/corda/membership/impl/registration/dynamic/mgm/MGMRegistrationContextValidator.kt:99]: I've just triggered a run using the changes, if that's enough: https://ci02.dev.r3.com/job/Corda5/job/corda5-upgrade-tests/job/PR-423/10/

Comment (yift-r3) [reply components/membership/registration-impl/src/main/kotlin/net/corda/membership/impl/registration/dynamic/mgm/MGMRegistrationContextValidator.kt:99]: Maybe add a unit test?

Comment (nikinagy) [reply components/membership/registration-impl/src/main/kotlin/net/corda/membership/impl/registration/dynamic/mgm/MGMRegistrationContextValidator.kt:99]: test added in [52670a6](https://github.com/corda/corda-runtime-os/pull/5573/commits/52670a6eb7fd74523eae870af52e0282c79adf9c)

File: testing/e2e-test-utilities/src/main/kotlin/net/corda/e2etest/utilities/MembershipUtils.kt (hunk 2 above)
Comment (nikinagy) [reply testing/e2e-test-utilities/src/main/kotlin/net/corda/e2etest/utilities/MembershipUtils.kt]: addressed in [52670a6](https://github.com/corda/corda-runtime-os/pull/5573/commits/52670a6eb7fd74523eae870af52e0282c79adf9c)

**Review by yift-r3 — APPROVED** [review]

**Pull request merged**